SignalServiceKit/src/Messages/Interactions/TSOutgoingMessage.m

```diff
@@ -203,23 +335,31 @@ - (void)saveWithTransaction:(YapDatabaseReadWriteTransaction *)transaction
     [super saveWithTransaction:transaction];
 }
 
-- (nullable NSString *)recipientIdentifier
+- (BOOL)hasSentToAnyRecipient
 {
-    return self.thread.contactIdentifier;
+    for (TSOutgoingMessageRecipientState *recipientState in self.recipientStateMap.allValues) {
+        if (recipientState.state == OWSOutgoingMessageRecipientStateSent) {
+            return YES;
+        }
+    }
+    return NO;
 }
 
 - (BOOL)shouldStartExpireTimer:(YapDatabaseReadTransaction *)transaction
 {
-    switch (self.messageState) {
-        case TSOutgoingMessageStateSentToService:
-            return self.isExpiringMessage;
-        case TSOutgoingMessageStateAttemptingOut:
-        case TSOutgoingMessageStateUnsent:
-            return NO;
-        case TSOutgoingMessageStateSent_OBSOLETE:
-        case TSOutgoingMessageStateDelivered_OBSOLETE:
-            OWSFail(@"%@ Obsolete message state.", self.logTag);
-            return self.isExpiringMessage;
+    // It's not clear if we should wait until _all_ recipients have reached "sent or later"
+    // (which could never occur if one group member is unregistered) or only wait until
+    // the first recipient has reached "sent or later" (which could cause partially delivered
+    // messages to expire).  For now, we'll do the latter.
+    //
+    // TODO: Revisit this decision.
+
+    if (!self.isExpiringMessage) {
+        return NO;
+    } else if (self.recipientStateMap.count < 1) {
```

Comment:
nit: I guess this `else if` is now redundant with `hasSentToAnyRecipient`

Reply:
Done.